drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,545 @@
+ *         (see @ref adc_emul_check_buffer_size)
+ * @return other error code returned by adc_context_wait_for_completion
+ */
+static int adc_emul_start_read(const struct device *dev,
+			       const struct adc_sequence *sequence)
+{
+	const struct adc_emul_config *config = dev->config;
+	struct adc_emul_data *data = dev->data;
+	int err;
+
+	if (sequence->resolution > ADC_EMUL_MAX_RESOLUTION ||
+	    sequence->resolution == 0) {
+		LOG_ERR("unsupported resolution %d", sequence->resolution);
+		return -ENOTSUP;
+	}
+
+	if (find_msb_set(sequence->channels) > config->num_channels) {
+		LOG_ERR("unsupported channels in mask: 0x%08x",
+			sequence->channels);
+		return -ENOTSUP;
+	}
+
+	err = adc_emul_check_buffer_size(dev, sequence);
+	if (err) {
+		LOG_ERR("buffer size too small");
+		return err;
+	}
+
+	data->res_mask = BIT_MASK(sequence->resolution);
+	data->buf = sequence->buffer;
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_emul_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	struct adc_emul_data *data = dev->data;
+	int err;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	err = adc_emul_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static int adc_emul_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	return adc_emul_read_async(dev, sequence, NULL);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buf = data->buf;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	if (repeat_sampling) {
+		data->buf = data->repeat_buf;
+	}
+}
+
+/**
+ * @brief Convert input voltage of ADC @p chan to raw output value
+ *
+ * @param data Internal data of ADC emulator
+ * @param chan ADC channel to sample
+ * @param result Raw output value
+ *
+ * @return 0 on success
+ * @return -EINVAL if failed to get reference voltage or unknown input is
+ *         selected
+ * @return other error code returned by custom function
+ */
+static int adc_emul_get_chan_value(struct adc_emul_data *data,
+				   unsigned int chan,
+				   adc_emul_res_t *result)
+{
+	struct adc_emul_chan_cfg *chan_cfg = &data->chan_cfg[chan];
+	uint32_t input_mV;
+	uint32_t ref_v;
+	uint64_t temp; /* Temporary 64 bit value prevent overflows */
+	int err;
+
+	/* Get input voltage */
+	switch (chan_cfg->input) {
+	case ADC_EMUL_CONST_VALUE:
+		input_mV = chan_cfg->const_value;
+		break;
+
+	case ADC_EMUL_CUSTOM_FUNC:
+		err = chan_cfg->func(data->dev, chan, chan_cfg->func_data,
```

Comment:
Don't you need to synchronize the access to `func` and `func_data` fields? The adc_emul API says the function can be changed asynchronously.

Reply:
I added `cfg_mtx` mutex which synchronize access to reference voltages and channels configuration.